feat: add migration to verify and fix embedded accumulated difficulty error

hansieodendaal opened the pull request:
Description
---
Added migration for accumulated difficulty, as we have an embedded accumulated difficulty error in some instances of the blockchain. It runs as a background task, and will stop when the last stored chain header has been processed.

Closes #7386.

Motivation and Context
---
See #7386.

How Has This Been Tested?
---
- System-level testing migrating the broken seed node's db from v5
- System-level testing a known good db from v1
- System-level testing a known good db from v5

What process can a PR reviewer use to test or verify this change?
---
- Code review
- System-level testing 

<!-- Checklist -->
<!-- 1. Is the title of your PR in the form that would make nice release notes? The title, excluding the conventional commit
tag, will be included exactly as is in the CHANGELOG, so please think about it carefully. -->


Breaking Changes
---

- [x] None
- [ ] Requires data directory on base node to be deleted
- [ ] Requires hard fork
- [ ] Other - Please specify

<!-- Does this include a breaking change? If so, include this line as a footer -->
<!-- BREAKING CHANGE: Description what the user should do, e.g. delete a database, resync the chain -->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Automatic, non-blocking background task that rebuilds accumulated difficulty data and can resume from the last processed height.
  * Public status and update endpoints to query and advance accumulated-data rebuild progress.

* **Migrations**
  * Migration version bumped; accumulated-data rebuild status is now recorded so upgrades can resume and report progress.

* **Bug Fixes**
  * Added additional known-good checkpoints to improve migration accuracy and robustness.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Commit:
Add migration for accumulated difficulty

Added migration for accumulated difficulty, as we have an embedded accumulated difficulty error in
some instances of the blockchain. It runs as a background task, and will stop when the last stored
chain header has been processed.

hansieodendaal commented:
As per technical discussion, this PR will be changed to not roll back the db but only recalculate the accumulated difficulty table as a background task.

Commit:
Merge remote-tracking branch 'upstream/development' into ho_acc_diff_issue

MCozhusheck commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -368,6 +369,83 @@ where B: BlockchainBackend
+                    Ok(Ok(current_status)) => {
+                        last_status = current_status;
+                    },
+                    Ok(Err(e)) => {
```
Maybe in a case this fails we should retry to recalculate difficulty again to make it more error prone.

SWvheerden replied:
I think this is fine as is, if we recalculate, what will have changed? Its kinda a big issue if this has broken.

hansieodendaal replied:
Agreed. If it does error here, the node will retry it the next time it is restarted from the last refreshed height.

stringhandler commented on base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs:
```diff
@@ -3995,3 +4115,76 @@ fn get_correct_accumulated_difficulty() -> Vec<(u64, U512)> {
+fn get_correct_accumulated_difficulty_v5() -> Vec<(u64, U512)> {
+    #[cfg(tari_target_network_mainnet)]
+    {
+        vec![
```
Didn't review properly, but this might be a problem. It might cause nodes to rollback to one of these heights

hansieodendaal replied:
It will not roll back anything, but it will trigger the accumulated difficulty to be recalculated from the last known good height onwards and be replaced in-place up to the last persisted chain header.

SWvheerden commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -2385,8 +2463,17 @@ fn get_vm_key_for_candidate_block<T: BlockchainBackend>(
+    get_vm_key_for_candidate_header(db, candidate_block.header.clone())
+}
+
+// this si tricky as we need to find the vm_key hash for the candidate block, but it might not be in the current chain
```
```suggestion
// this is tricky as we need to find the vm_key hash for the candidate block, but it might not be in the current chain
```

SWvheerden commented on base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs:
```diff
@@ -3995,3 +4115,76 @@ fn get_correct_accumulated_difficulty() -> Vec<(u64, U512)> {
     vec![]
 }
+
+fn get_correct_accumulated_difficulty_v5() -> Vec<(u64, U512)> {
```
why not merge the two?

SWvheerden commented on base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs:
```diff
@@ -3995,3 +4115,76 @@ fn get_correct_accumulated_difficulty() -> Vec<(u64, U512)> {
+        vec![
+            // We have known bad values at the next height, so this one can be a good reference point
+            (
+                15224,
```
we dont need this fine grained loop, rough estimates will be fine as we will fix these in the background. And it should be quick.

SWvheerden reviewed (commented):
few comments here

Commit:
review comments

Commit:
Merge remote-tracking branch 'upstream/development' into ho_acc_diff_issue

SWvheerden commented on base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs:
```diff
@@ -3945,7 +4064,7 @@ pub struct OldChainTipData {
 }
 
-fn get_correct_accumulated_difficulty() -> Vec<(u64, U512)> {
+fn get_correct_accumulated_difficulty_v1_and_v5() -> Vec<(u64, U512)> {
```
```suggestion
fn get_correct_accumulated_difficulty() -> Vec<(u64, U512)> {
```

SWvheerden commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -2859,6 +2946,39 @@ fn process_payref_for_height<B: BlockchainBackend>(
+) -> Result<AccumulatedDataRebuildStatus, ChainStorageError> {
+    debug!(target: LOG_TARGET, "[AccData] Processing accumulated data rebuilding for height {}", height);
+
+    let write_txn = db
```
```suggestion
    let write_lock = db
```

SWvheerden replied:
this is not a write txn, its a write_lock on the db.

SWvheerden commented on base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs:
```diff
@@ -3884,7 +3909,101 @@ fn run_migrations(db: &mut LMDBDatabase) -> Result<(), ChainStorageError> {
+        // good accumulated difficulty
+        if migrate_from_version == 5 {
+            let mut known_good_difficulties = get_correct_accumulated_difficulty_v1_and_v5();
+            known_good_difficulties.sort_by(|a, b| a.0.cmp(&b.0));
```
```suggestion
```

SWvheerden replied:
its already a sorted unified list

Commit:
review comments

hansieodendaal commented on base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs:
```diff
@@ -2879,6 +2893,39 @@ impl BlockchainBackend for LMDBDatabase {
+        write_txn.commit()?;
+
+        Ok(status)
+    }
```
All consistency gaurantees are upheld in the calling function, `fn process_accumulated_data_for_height<B: BlockchainBackend>(` as this code is only used for a once-off migration. This code only runs if we have existing `ChainHeder` data in the lmdb, which includes the `BlockHeaderAccumulatedData`.

SWvheerden committed:
Merge branch 'development' into ho_acc_diff_issue

SWvheerden committed:
Merge branch 'development' into ho_acc_diff_issue

Commit:
Merge remote-tracking branch 'upstream/development' into ho_acc_diff_issue

# Conflicts:
#	base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs
#	base_layer/core/src/chain_storage/lmdb_db/mod.rs

Commit:
Fix compile errors, not using nightly

- `cargo clippy --workspace --exclude tari_libtor --all-targets --all-features -- -D warnings`
- `cargo lints clippy --workspace --exclude tari_libtor --all-targets --all-features`

Commit:
cargo machete, unused features

Commit:
Merge branch 'ho_compile_errors' into ho_acc_diff_issue

Commit:
merge
